### DISCUSSION TITLE
ExportSolidity Public Input Size Problem

### DISCUSSION THREAD

**Discussion by SherLzp** [Consensys/gnark#644] [Q&A]
Hi sir, @ivokub 

```go
func exportMultiECDSACircuit(size int64) {
	var circuit ecdsa.MultiEcdsaCircuit[emulated.Secp256r1Fp, emulated.Secp256r1Fr]
	circuit.Ecdsas = make([]ecdsa.EcdsaCircuit[emulated.Secp256r1Fp, emulated.Secp256r1Fr], size)
	oR1cs, err := frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder, &circuit, frontend.IgnoreUnconstrainedInputs())
	if err != nil {
		panic(err)
	}

	pk, vk, err := groth16.Setup(oR1cs)
	if err != nil {
		panic(err)
	}
	{
		f, err := os.Create("./keys/ecdsa/kontos-ecdsa-" + fmt.Sprint(size) + ".vk")
		if err != nil {
			panic(err)
		}
		_, err = vk.WriteRawTo(f)
		if err != nil {
			panic(err)
		}
	}
	{
		f, err := os.Create("./keys/ecdsa/kontos-ecdsa-" + fmt.Sprint(size) + ".pk")
		if err != nil {
			panic(err)
		}
		_, err = pk.WriteRawTo(f)
		if err != nil {
			panic(err)
		}
	}

	{
		f, err := os.Create("./keys/ecdsa/KontosECDSAVerifier-" + fmt.Sprint(size) + ".sol")
		if err != nil {
			panic(err)
		}
		err = vk.ExportSolidity(f)
		if err != nil {
			panic(err)
		}
	}
}

```

I want to verify ECDSA circuit with solidity, then I try the above method, why it has 21 public inputs?

```js
function verifyProof(
    uint256[2] memory a,
    uint256[2][2] memory b,
    uint256[2] memory c,
    uint256[21] calldata input
) public view returns (bool r)
```

**Comment by ivokub** [comment]
Can you give the circuit type definition (`type Circuit struct ...`)? If you provide as an input the public key, message, signature, then this amounts to approximately this amount of public inputs (key 8 vars, message 4 vars, signature 9 vars).

This can theoretically be reduced, but it depends on your use-case. E.g. you can provide signature as a private input (statement "there is a valid signature for message by public key") or you can provide the public key as an address (but then have to hash public key into address in-circuit).

**Reply by SherLzp** [reply]
> Can you give the circuit type definition (`type Circuit struct ...`)? If you provide as an input the public key, message, signature, then this amounts to approximately this amount of public inputs (key 8 vars, message 4 vars, signature 9 vars).
> 
> This can theoretically be reduced, but it depends on your use-case. E.g. you can provide signature as a private input (statement "there is a valid signature for message by public key") or you can provide the public key as an address (but then have to hash public key into address in-circuit).

yes, that's the same with this one:

```go
type EcdsaCircuit[T, S emulated.FieldParams] struct {
	Sig Signature[S]
	Msg emulated.Element[S]
	Pub PublicKey[T, S]
}

func (c *EcdsaCircuit[T, S]) Define(api frontend.API) error {
	c.Pub.Verify(api, sw_emulated.GetCurveParams[T](), &c.Msg, &c.Sig)
	return nil
}
```

So we need to convert uint256(solidity) to `Limbs` in solidity code? is there any example for that?

**Reply by SherLzp** [reply]
And why Sig has 9 vars? I think that's should be 8?

```go
// Signature represents the signature for some message.
type Signature[Scalar emulated.FieldParams] struct {
	R, S emulated.Element[Scalar]
}
```

**Reply by SherLzp** [reply]
Hi sir, @ivokub 
It seems whenever I use emulated, it will always have one more public input. Is that scalar field of the verify curve?

**Reply by ivokub** [reply]
Regarding the signature, I was mistaken. I was thinking about the Ecrecover precompile which takes additional input (v value). For plain ECDSA signature it should be 8 inputs.

When we use field emulation then we use augmented proof system which has one more value in the proof. I'm not sure why it is part of the input though, I didn't review the implementation. I think it is expected but have to double-check. The augmented proof system allows to have order of magnitude faster prover.

**Reply by hussein-aitlahcen** [reply]
I am facing the same issue. I'm not using any emulated element as part of the public inputs though (only `frontend.Variable`).
```go
	Root             [2]frontend.Variable `gnark:",public"`
	SecondRoot [2]frontend.Variable `gnark:",public"`
	Data             [2]frontend.Variable `gnark:",public"`
```
I expected the inputs to be 6 but it's 7, and I'm not sure where to extract the 7th input from. When marshalling the public witness, I only get 6 values:

`0x00000006000000000000000600000000000000000000000000000000d9a17d56267ffefdef74c37aa87bf09500000000000000000000000000000000decd38b6540daeacd4fdfe40418f6c8900000000000000000000000000000000d9a17d56267ffefdef74c37aa87bf09500000000000000000000000000000000decd38b6540daeacd4fdfe40418f6c89263cb8e7b41d2ef3a4ee22f8bc8bfd54d3e93372fbe8e26cf473c8c5af1d5b7128dcc7cf00b1a8f58ab328b18a7ca34e4c45a774f2d8cd926f0ac0172633d3ac`

`vk.NbPublicInput()` does return `7` as well... :thinking:

**Reply by hussein-aitlahcen** [reply]
Ah, seems like I found what you mention @ivokub https://github.com/ConsenSys/gnark/blob/develop/backend/groth16/bn254/setup.go#L101, thanks for pointing :)

We need to append the commitment hash as how the go verifier is doing https://github.com/ConsenSys/gnark/blob/develop/backend/groth16/bn254/verify.go#L80
https://github.com/ConsenSys/gnark/blob/develop/backend/groth16/bn254/commitment.go#L27

Also, look like the proof commitment is also added here https://github.com/ConsenSys/gnark/blob/develop/backend/groth16/bn254/verify.go#L92

The solidity code is broken then?

I managed to make the verifier working by adding the extra addition, which involve adding 2 public inputs... Which mean we now have 3 extra inputs (18k gas). I'm not sure the prover perf benefit are worth in certain scenario (18k gas on EVM is not cheap).

**Reply by ivokub** [reply]
As far as I know then @Tabaie is right now actively working on verifiers (both Groth16 and PLONK) in Solidity (https://github.com/ConsenSys/gnark/pull/626 for example). I would give a few weeks until the implementations start to settle.

**Reply by ivokub** [reply]
> I managed to make the verifier working by adding the extra addition, which involve adding 2 public inputs... Which mean we now have 3 extra inputs (18k gas). I'm not sure the prover perf benefit are worth in certain scenario (18k gas on EVM is not cheap).

That is good knowledge. I actually have some idea of having a flag to prevent the commitment API when the verifier is constrained (e.g. want to really really optimise for gas). But will get back to these when we have done the optimised Solidity verifiers.

**Reply by hussein-aitlahcen** [reply]
> As far as I know then @Tabaie is right now actively working on verifiers (both Groth16 and PLONK) in Solidity (#626 for example). I would give a few weeks until the implementations start to settle.

I'm impressed you guys already tackle every issue we face haha. Thanks for pointing again!

**Reply by ivokub** [reply]
Tracking in https://github.com/ConsenSys/gnark/issues/652.

**Reply by Tabaie** [reply]
Credit where it is due, @ThomasPiellard has done most of it. Is Groth16 commitments in Solidity a priority for you guys?

**Reply by SherLzp** [reply]
> Credit where it is due, @ThomasPiellard has done most of it. Is Groth16 commitments in Solidity a priority for you guys?

Yes! Reduce the gas on-chain is the most important part for verifying.

**Reply by zhengjunhe** [reply]
@ivokub could you pls tell me how to fill the extra 32 byte info

**Reply by ivokub** [reply]
Groth16 Solidity verifier still does not support verifying the commitments. It is right now blocked by not having hash-to-field in Solidity smart contract verifier.

PLONK verifier should work though. The prover is slower, but at least we have smart contract verifier for it.

**Reply by danivilardell** [reply]
Hi @ivokub, is there any plan to implement commitment verification on Solidity verifier some time soon?

**Reply by ivokub** [reply]
Hi, gnark latest release v0.10 supports commitment verification in Solidity Groth16 verifier. However, for it to work you have to provide `sha2-256` as the hash function during proving time:
```go
groth16.Prove(r1cs, pk, witness, backend.WithProverHashToFieldFunction(sha256.New()))
```
(keep in mind the `backend.WithProverHashToFieldFunction` option. And for formatting the calldata see the [example](https://github.com/Consensys/gnark-solidity-checker/blob/main/cmd/templates.go#L101-L111)

**Comment by SherLzp** [comment]
Hi sir, @ivokub  
@gbotrel @yelhousni 
One more question here.

I mark here as public:
https://github.com/ConsenSys/gnark/blob/master/std/signature/ecdsa/ecdsa_test.go#L18

I add one line here:
```go
assert.ProverSucceeded(&circuit, &witness, test.WithCurves(ecc.BN254), test.WithBackends(backend.GROTH16))
```

to this part: https://github.com/ConsenSys/gnark/blob/master/std/signature/ecdsa/ecdsa_test.go#L68

However, it will occur the following problems:

```go
panic: runtime error: slice bounds out of range [:768] with capacity 704 [recovered]
	panic: runtime error: slice bounds out of range [:768] with capacity 704
```

**Reply by ivokub** [reply]
May be a bug. Will have a look in the issue later.

**Reply by SherLzp** [reply]
Sir, @ivokub 
Find the bug here: https://github.com/ConsenSys/gnark/blob/develop/constraint/commitment.go#L41
```go
func (i *Commitment) SerializeCommitment(privateCommitment []byte, publicCommitted []*big.Int, fieldByteLen int) []byte {

	res := make([]byte, len(privateCommitment)+len(publicCommitted)*fieldByteLen)
	copy(res, privateCommitment)

	offset := len(privateCommitment)
	for j, inJ := range publicCommitted {
		offset += j * fieldByteLen
		inJ.FillBytes(res[offset : offset+fieldByteLen])
	}

	return res
}
```

I think it should be changed to:
```go
func (i *Commitment) SerializeCommitment(privateCommitment []byte, publicCommitted []*big.Int, fieldByteLen int) []byte {

	res := make([]byte, len(privateCommitment)+len(publicCommitted)*fieldByteLen)
	copy(res, privateCommitment)

	offset := len(privateCommitment)
	for _, inJ := range publicCommitted {
		inJ.FillBytes(res[offset : offset+fieldByteLen])
		offset += fieldByteLen
	}

	return res
}
```

then it works.

I make a PR here: #651

**Reply by SherLzp** [reply]
However, I met another bug here:

When I just use the following code to prove/verify, it works:
```go
pk, vk, err := groth16.Setup(oR1cs)
assert.NoError(err)
witnessData, err := frontend.NewWitness(&witness, ecc.BN254.ScalarField())
assert.NoError(err)

proof, err := groth16.Prove(oR1cs, pk, witnessData)
assert.NoError(err)

publicWitnessData, err := frontend.NewWitness(&witness, ecc.BN254.ScalarField())
assert.NoError(err)
err = groth16.Verify(proof, vk, publicWitnessData)
assert.NoError(err)
```

However, when I write pk/vk to file and read it again, it will fail, I think that's the bug for marshal/unmarshal:
```go
Error:      	Received unexpected error:
        	   constraint #309736 is not satisfied: [div] 59521/0 == <unsolved>
        	   r1cs.(*builder).DivUnchecked
        	    api.go:257
        	   rangecheck.(*commitChecker).commit.func2
        	    rangecheck_commit.go:99
```

The keys are not match:
```go
	vkRead := groth16.NewVerifyingKey(ecc.BN254)
	readVkFile, err := os.Open("xx.vk")
	if err != nil {
		panic(err)
	}
	_, err = vkRead.ReadFrom(readVkFile)
	if err != nil {
		panic(err)
	}

	isEqual := reflect.DeepEqual(pk, pkRead)
	if !isEqual {
		panic("unequal key")
	}
```

**Reply by hussein-aitlahcen** [reply]
Proving/verifying keys are missing commitment info when serialized, see https://github.com/ConsenSys/gnark/issues/580

**Reply by SherLzp** [reply]
> Proving/verifying keys are missing commitment info when serialized, see #580

Thank you! Got it. And already fixed it on my branch.
